File: scripts/species_richness_analysis.py
#!/usr/bin/env python3
"""
Species Richness Analysis for Annotation Reports
Analysiert die Artenvielfalt in allen Annotation-Dateien
"""

import pandas as pd
import os
from pathlib import Path
from scipy.stats import mannwhitneyu

def calculate_species_richness(file_path, file_name):
    """Berechnet die Species Richness für eine Datei"""
    # Daten einlesen
    data = pd.read_csv(file_path)
    
    # Eindeutige Arten zählen
    unique_species = data['label_name'].nunique()
    
    # Gesamtanzahl der Beobachtungen
    total_observations = len(data)
    
    # Arten mit Häufigkeit
    species_counts = data.groupby(['label_name', 'label_hierarchy']).size().reset_index(name='count')
    species_counts = species_counts.sort_values('count', ascending=False)
    
    return {
        'file_name': file_name,
        'species_richness': unique_species,
        'total_observations': total_observations,
        'species_counts': species_counts
    }

def main():
    # Alle CSV-Dateien im Annotation_reports Ordner finden
    annotation_dir = Path("Annotation_reports")
    csv_files = list(annotation_dir.glob("*.csv"))
    
    # Species Richness für alle Dateien berechnen
    results_list = []
    for file in csv_files:
        file_name = file.name
        print(f"Analysiere: {file_name}")
        result = calculate_species_richness(file, file_name)
        results_list.append(result)
    
    # Zusammenfassung erstellen
    summary_data = []
    for res in results_list:
        summary_data.append({
            'file_name': res['file_name'],
            'species_richness': res['species_richness'],
            'total_observations': res['total_observations']
        })
    
    summary_df = pd.DataFrame(summary_data)
    
    # Standort und Köder aus Dateinamen extrahieren
    summary_df['location'] = summary_df['file_name'].apply(
        lambda x: 'Utumbi' if 'utumbi' in x.lower() else 'Milimani'
    )
    summary_df['date'] = summary_df['file_name'].str.extract(r'^(\d+)-')[0]
    summary_df['bait_info'] = summary_df['file_name'].str.extract(r'-c\d+-(.+)-ganz\.csv$')[0]
    summary_df['observations_per_species'] = (
        summary_df['total_observations'] / summary_df['species_richness']
    ).round(2)
    
    # Sortieren nach Species Richness
    summary_df = summary_df.sort_values('species_richness', ascending=False)
    
    # Ergebnisse ausgeben
    print("\n=== SPECIES RICHNESS VERGLEICH ===\n")
    print(summary_df.to_string(index=False))
    
    # Detaillierte Artenlisten für jede Datei
    print("\n\n=== DETAILLIERTE ARTENLISTEN ===\n")
    for res in results_list:
        print(f"\n--- {res['file_name']} ---")
        print(f"Species Richness: {res['species_richness']}")
        print(f"Gesamtbeobachtungen: {res['total_observations']}\n")
        print("Top 10 häufigste Arten:")
        print(res['species_counts'].head(10).to_string(index=False))
        print()
    
    # Ergebnisse in CSV speichern
    output_file = "results/species_richness_comparison.csv"
    os.makedirs("results", exist_ok=True)
    summary_df.to_csv(output_file, index=False)
    print(f"\nErgebnisse gespeichert in: {output_file}")
    
    # Zusätzliche detaillierte Artenliste mit allen Arten pro Datei
    detailed_species_list = []
    for res in results_list:
        species_df = res['species_counts'].copy()
        species_df['file_name'] = res['file_name']
        detailed_species_list.append(species_df)
    
    detailed_df = pd.concat(detailed_species_list, ignore_index=True)
    detailed_output_file = "results/species_richness_detailed.csv"
    detailed_df.to_csv(detailed_output_file, index=False)
    print(f"Detaillierte Artenliste gespeichert in: {detailed_output_file}")
    
    # ===== STATISTISCHE SIGNIFIKANZTESTS =====
    
    print("\n\n===== STATISTISCHE SIGNIFIKANZTESTS =====\n")
    
    significance_results = []
    
    # Vergleich zwischen Standorten (Milimani vs Utumbi)
    milimani_richness = summary_df[summary_df['location'] == 'Milimani']['species_richness'].values
    utumbi_richness = summary_df[summary_df['location'] == 'Utumbi']['species_richness'].values
    
    if len(milimani_richness) > 0 and len(utumbi_richness) > 0:
        u_stat, p_value = mannwhitneyu(milimani_richness, utumbi_richness)
        print(f"Milimani vs Utumbi (Species Richness):")
        print(f"  n_Milimani: {len(milimani_richness)}")
        print(f"  n_Utumbi: {len(utumbi_richness)}")
        print(f"  Test: Mann-Whitney U")
        print(f"  Statistic: {u_stat:.4f}")
        print(f"  p-value: {p_value:.6f}")
        print(f"  Significant (α=0.05): {'Yes' if p_value < 0.05 else 'No'}\n")
        
        significance_results.append({
            'comparison': 'Milimani vs Utumbi',
            'n_group1': len(milimani_richness),
            'n_group2': len(utumbi_richness),
            'test': 'Mann-Whitney U',
            'statistic': f"{u_stat:.4f}",
            'p_value': f"{p_value:.6f}",
            'significant_alpha_005': 'Yes' if p_value < 0.05 else 'No',
            'mean_group1': f"{milimani_richness.mean():.2f}",
            'mean_group2': f"{utumbi_richness.mean():.2f}"
        })
    
    # Speichere Signifikanz-Ergebnisse
    stats_df = pd.DataFrame(significance_results)
    stats_output_file = "results/species_richness_statistical_tests.csv"
    stats_df.to_csv(stats_output_file, index=False)
    print(f"Signifikanz-Tests gespeichert in: {stats_output_file}")

if __name__ == "__main__":
    main()

File: scripts/species_richness_analysis.R
# Species Richness Analysis for Annotation Reports
# Analysiert die Artenvielfalt in allen Annotation-Dateien

library(dplyr)
library(tidyr)

# Funktion zur Berechnung der Species Richness für eine Datei
calculate_species_richness <- function(file_path, file_name) {
  # Daten einlesen
  data <- read.csv(file_path, stringsAsFactors = FALSE)
  
  # Eindeutige Arten zählen
  unique_species <- data %>%
    select(label_name) %>%
    distinct() %>%
    nrow()
  
  # Gesamtanzahl der Beobachtungen
  total_observations <- nrow(data)
  
  # Arten mit Häufigkeit
  species_counts <- data %>%
    group_by(label_name, label_hierarchy) %>%
    summarise(count = n(), .groups = "drop") %>%
    arrange(desc(count))
  
  # Ergebnis zurückgeben
  list(
    file_name = file_name,
    species_richness = unique_species,
    total_observations = total_observations,
    species_counts = species_counts
  )
}

# Alle CSV-Dateien im Annotation_reports Ordner finden
annotation_dir <- "Annotation_reports"
csv_files <- list.files(annotation_dir, pattern = "\\.csv$", full.names = TRUE)

# Species Richness für alle Dateien berechnen
results_list <- lapply(csv_files, function(file) {
  file_name <- basename(file)
  cat("Analysiere:", file_name, "\n")
  calculate_species_richness(file, file_name)
})

# Zusammenfassung erstellen
summary_df <- do.call(rbind, lapply(results_list, function(res) {
  data.frame(
    file_name = res$file_name,
    species_richness = res$species_richness,
    total_observations = res$total_observations,
    stringsAsFactors = FALSE
  )
}))

# Standort und Köder aus Dateinamen extrahieren
summary_df <- summary_df %>%
  mutate(
    # Dateiname aufteilen
    location = ifelse(grepl("utumbi", tolower(file_name)), "Utumbi", "Milimani"),
    date = sub("^(\\d+)-.*", "\\1", file_name),
    bait_info = sub(".*-c\\d+-(.+)-ganz\\.csv$", "\\1", file_name),
    # Shannon Diversity Index könnte später berechnet werden
    observations_per_species = round(total_observations / species_richness, 2)
  )

# Sortieren nach Species Richness
summary_df <- summary_df %>%
  arrange(desc(species_richness))

# Ergebnisse ausgeben
cat("\n=== SPECIES RICHNESS VERGLEICH ===\n\n")
print(summary_df)

# Detaillierte Artenlisten für jede Datei
cat("\n\n=== DETAILLIERTE ARTENLISTEN ===\n\n")
for (res in results_list) {
  cat("\n--- ", res$file_name, " ---\n")
  cat("Species Richness:", res$species_richness, "\n")
  cat("Gesamtbeobachtungen:", res$total_observations, "\n\n")
  cat("Top 10 häufigste Arten:\n")
  print(head(res$species_counts, 10))
  cat("\n")
}

# Ergebnisse in CSV speichern
output_file <- "results/species_richness_comparison.csv"
write.csv(summary_df, output_file, row.names = FALSE)
cat("\nErgebnisse gespeichert in:", output_file, "\n")

# Zusätzliche detaillierte Artenliste mit allen Arten pro Datei
detailed_species_list <- do.call(rbind, lapply(results_list, function(res) {
  res$species_counts %>%
    mutate(file_name = res$file_name)
}))

detailed_output_file <- "results/species_richness_detailed.csv"
write.csv(detailed_species_list, detailed_output_file, row.names = FALSE)
cat("Detaillierte Artenliste gespeichert in:", detailed_output_file, "\n")

# ===== STATISTISCHE SIGNIFIKANZTESTS =====

cat("\n\n===== STATISTISCHE SIGNIFIKANZTESTS =====\n\n")

# Vergleich zwischen Standorten (Milimani vs Utumbi)
milimani_richness <- summary_df %>%
  filter(location == "Milimani") %>%
  pull(species_richness)

utumbi_richness <- summary_df %>%
  filter(location == "Utumbi") %>%
  pull(species_richness)

significance_results <- data.frame()

if (length(milimani_richness) > 0 && length(utumbi_richness) > 0) {
  test_result <- wilcox.test(milimani_richness, utumbi_richness)
  cat("Milimani vs Utumbi (Species Richness):\n")
  cat("  n_Milimani:", length(milimani_richness), "\n")
  cat("  n_Utumbi:", length(utumbi_richness), "\n")
  cat("  Test: Mann-Whitney U\n")
  cat("  Statistic:", round(test_result$statistic, 4), "\n")
  cat("  p-value:", round(test_result$p.value, 6), "\n")
  cat("  Significant (α=0.05):", ifelse(test_result$p.value < 0.05, "Yes", "No"), "\n\n")
  
  significance_results <- rbind(significance_results, data.frame(
    comparison = "Milimani vs Utumbi",
    n_group1 = length(milimani_richness),
    n_group2 = length(utumbi_richness),
    test = "Mann-Whitney U",
    statistic = round(test_result$statistic, 4),
    p_value = round(test_result$p.value, 6),
    significant_alpha_005 = ifelse(test_result$p.value < 0.05, "Yes", "No"),
    mean_group1 = round(mean(milimani_richness), 2),
    mean_group2 = round(mean(utumbi_richness), 2)
  ))
}

# Speichere Signifikanz-Ergebnisse
write.csv(significance_results, "results/species_richness_statistical_tests.csv", row.names = FALSE)
cat("Signifikanz-Tests gespeichert in: results/species_richness_statistical_tests.csv\n")
